presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java

```diff
@@ -37,25 +41,43 @@
      */
     private final List<Symbol> correlation;
 
+    /**
+     * Assignments that operate on subqueries (e.g: "expr IN (SELECT 1)").
+     * It might be a direct mapping to a subquery symbol if subquery is scalar.
+     */
+    private final Map<Symbol, Expression> subqueryAssignments;
+
     @JsonCreator
     public ApplyNode(
             @JsonProperty("id") PlanNodeId id,
             @JsonProperty("input") PlanNode input,
             @JsonProperty("subquery") PlanNode subquery,
+            @JsonProperty("subqueryAssignments") Map<Symbol, Expression> subqueryAssignments,
             @JsonProperty("correlation") List<Symbol> correlation)
     {
         super(id);
         requireNonNull(input, "input is null");
         requireNonNull(subquery, "right is null");
+        requireNonNull(subqueryAssignments, "assignments is null");
         requireNonNull(correlation, "correlation is null");
 
         checkArgument(input.getOutputSymbols().containsAll(correlation), "Input does not contain symbols from correlation");
 
         this.input = input;
         this.subquery = subquery;
+        this.subqueryAssignments = subqueryAssignments;
         this.correlation = ImmutableList.copyOf(correlation);
     }
 
+    /**
+     * @return true when subquery is scalar and it's output symbols are directly mapped to ApplyNode output symbols
+     */
+    public boolean isResolvedScalarSubquery()
```

Comment:
`ScalarQueryUtil` returns true if plan node returns single node. `ApplyNode` can have scalar subquery, but not be scalar, so it doesn't quite fit there.